Firestore/core/src/util/executor_libdispatch.mm

```diff
@@ -243,91 +113,184 @@ void MarkDone() {
+  // to truly cancel work after it's been dispatched, libdispatch is guaranteed
+  // to outlive the executor, and it's possible for work to be invoked by
+  // libdispatch after the executor is destroyed. The Executor only stores an
+  // observer pointer to the operation.
+  Task* task = nullptr;
+  TimePoint target_time = MakeTargetTime(delay);
+  Id id = 0;
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    id = NextIdLocked();
+    task = Task::Create(this, target_time, tag, id, std::move(operation));
+    async_tasks_.insert(task);
+    schedule_[id] = task;
+  }
+
+  dispatch_after_f(delay_ns, dispatch_queue_, task, InvokeAsync);
+
+  return DelayedOperation(this, id);
+}
+
+void ExecutorLibdispatch::Complete(Task* task) {
+  bool should_release = false;
+  {
+    TASK_TRACE("Executor::Complete %s task %s", this, task);
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    auto found = async_tasks_.find(task);
+    if (found != async_tasks_.end()) {
+      should_release = true;
+      async_tasks_.erase(found);
+
+      if (!task->is_immediate()) {
+        schedule_.erase(task->id());
+      }
+    }
+  }
+
+  if (should_release) {
+    task->Release();
+  }
+}
+
+void ExecutorLibdispatch::Cancel(Id operation_id) {
+  Task* found_task = nullptr;
+  {
+    TASK_TRACE("Executor::Cancel %s task %s", this, task);
+    std::lock_guard<std::mutex> lock(mutex_);
+
     // `time_slot` might have been destroyed by the time cancellation function
     // runs, in which case it's guaranteed to have been removed from the
     // `schedule_`. If the `time_slot_id` refers to a slot that has been
     // removed, the call to `RemoveFromSchedule` will be a no-op.
-    RemoveFromSchedule(time_slot_id);
-  }};
-}
-
-void ExecutorLibdispatch::RemoveFromSchedule(TimeSlotId to_remove) {
-  RunSynchronized(this, [this, to_remove] {
-    const auto found = schedule_.find(to_remove);
+    const auto found = schedule_.find(operation_id);
 
     // It's possible for the operation to be missing if libdispatch gets to run
     // it after it was force-run, for example.
     if (found != schedule_.end()) {
-      found->second->MarkDone();
+      found_task = found->second;
+      async_tasks_.erase(found_task);
       schedule_.erase(found);
     }
-  });
+  }
+
+  if (found_task) {
+    found_task->Cancel();
+    found_task->Release();
+  }
 }
 
-// Test-only methods
+void ExecutorLibdispatch::InvokeAsync(void* raw_task) {
+  auto task = static_cast<Task*>(raw_task);
+  task->Execute();
+}
 
-bool ExecutorLibdispatch::IsScheduled(const Tag tag) const {
-  bool result = false;
-  RunSynchronized(this, [this, tag, &result] {
-    result = std::any_of(schedule_.begin(), schedule_.end(),
-                         [&tag](const ScheduleEntry& operation) {
-                           return *operation.second == tag;
-                         });
-  });
-  return result;
+void ExecutorLibdispatch::InvokeSync(void* raw_task) {
```

Comment:
These two functions have the same implementation. Do you think it's worthwhile keeping them separate (e.g. to be more future-proof)?

Reply:
I kept them separate because stack traces look better--it's purely cosmetic. I've added a comment so that a future do-gooder doesn't undo this.